File: Oppg2.py
#Ny pythonfil
#python Oppg2.py

from threading import Thread, Lock

times = 500;
i = 0

def Inkrementer(lock):
  global i
  for _ in range(times):
    lock.acquire()
    i += 2
    lock.release()
  
def Dekrementer(lock):
  global i
  for _ in range(times):
    lock.acquire()
    i -= 1
    lock.release()
  
def main():
  lock=Lock()
  someThread = Thread(target = Inkrementer, args = (lock,),)
  someThread2 = Thread(target = Dekrementer, args = (lock,),)
  someThread.start()
  someThread2.start()
  someThread.join()
  someThread2.join()
  print(i)

main()

File: Oppg2.c
#include <pthread.h>
#include <stdio.h>

int i = 0;
int N = 1000000;

// Note the return type: void*
void* Inkrementer(){
    for (int j = 0; j < N; j++)
    	i++;
    return NULL;
}

void* Dekrementer(){
    for (int j = 0; j < N; j++)
    	i--;
    return NULL;
}



int main(){
    pthread_t someThread;
    pthread_t someThread2;
    pthread_create(&someThread, NULL, Inkrementer, NULL);
    pthread_create(&someThread2, NULL, Dekrementer, NULL);
    
    pthread_join(someThread, NULL);
    pthread_join(someThread2, NULL);
    printf("%d\n", i);
    return 0;
    
}

